[Cikguseven] iP

Cikguseven opened the pull request:
### Oscar

> "I understand that, without my agreement, Alpine F1 have put out a press release late this afternoon that I am driving for them next year. This is wrong and I have not signed a contract with Alpine for 2023. I will not be driving for Alpine next year." - Oscar Piastri ([source](https://twitter.com/OscarPiastri/status/1554527452231262210))

Oscar is a chatbot named after F1 Driver Oscar Piastri. Oscar is, 

- text-based 📄 
- user friendly 😃 
- SUPER DUPER FAST to use 🚀 

To use Oscar, all you need to do is ,
1. download the "oscar-1.0.0.jar" jar file from [here](https://github.com/Cikguseven/ip/releases/tag/Week-3).
2. save it to any folder.
3. navigate to that folder.
4. open the terminal in that folder.
5. run the command `java -jar oscar-1.0.0.jar`.
6. enjoy!

And it is 100% **FREE**!

Features:

- [X]  Managing tasks
- [X] Managing deadlines
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
    new Oscar(FILE_PATH).run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Cikguseven committed:
Initial skeleton version of Oscar chatbot

Cikguseven committed:
Skeletal version of Oscar with echo

Cikguseven committed:
Oscar can now store user text

Cikguseven committed:
Completed Level 3 with mark and unmark features

Cikguseven committed:
Completed Level 4.

Cikguseven committed:
Completed A-TextUiTesting

Cikguseven committed:
Completed Level 5.

Cikguseven committed:
Completed Level 6.

Cikguseven committed:
Fixed exception handling in Duke.java

Cikguseven committed:
Added enums for commands

Cikguseven committed:
Renamed Duke.java to Oscar.java

Cikguseven committed:
Renamed DukeException.java to OscarException.java

Cikguseven committed:
Reformat case statements.

Cikguseven committed:
Add Level 7 features

Cikguseven committed:
Merge branch 'branch-Level-7'

Cikguseven committed:
Initial implementation of Level 8

Cikguseven committed:
Add initial version of date time support.

Cikguseven committed:
Add Level 8

Cikguseven committed:
Merge branch 'branch-Level-8'

Cikguseven committed:
Modify project to use more OOP

Cikguseven committed:
Merge branch 'branch-A-MoreOOP'

Cikguseven committed:
Organize classes into packages

Cikguseven committed:
Merge branch 'branch-A-Packages'

Cikguseven committed:
Add files for tracking

Cikguseven committed:
Format imports

Cikguseven committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Cikguseven committed:
Add Gradle

Cikguseven committed:
Add Junit Tests

Cikguseven committed:
Merge branch 'branch-A-JUnit'

Cikguseven committed:
Fix method signature of commands

Cikguseven committed:
Create JAR File

Cikguseven committed:
Add Find

Cikguseven committed:
Merge branch 'branch-Level-9'

Cikguseven committed:
Follow the coding standard

Cikguseven committed:
Merge branch 'branch-A-CodingStandard'

Cikguseven committed:
Add JavaDoc

Cikguseven committed:
Add parallel increments

Cikguseven committed:
Add checkstyle files

Cikguseven committed:
Fix checkstyle

Cikguseven committed:
Merge branch 'branch-A-CheckStyle'

Cikguseven committed:
Update README.md

Cikguseven committed:
Update README.md

Cikguseven committed:
Modify build.gradle to support JavaFX

0xteet0 commented on src/main/java/oscar/task/Task.java:
```diff
@@ -0,0 +1,67 @@
+public abstract class Task implements Serializable {
+    private final String description;
+    private final String type;
+    private boolean isDone;
```
Great use of coding standard here when naming boolean variables!

0xteet0 commented on src/main/java/oscar/essential/Parser.java:
```diff
@@ -0,0 +1,73 @@
+        String details = fullCommand.length() > command.length() ? splits[1] : "";
+        switch (command) {
+        // Exit programme if user enters "bye" command
+        case "bye":
```
Nice! Great to see that you have abided the indentation for switch statements, avoiding the common indentation mistake!

0xteet0 commented on src/main/java/oscar/essential/TaskList.java:
```diff
@@ -0,0 +1,164 @@
+        this.taskList = load(stream);
+    }
+
+    /**
```
Good job on the JavaDoc comments! Clear and concise!

0xteet0 commented on src/main/java/oscar/task/Deadline.java:
```diff
@@ -0,0 +1,33 @@
+     * @return Information of deadline.
+     */
+    @Override
+    public String toString() {
```
Nice to see that line length < 120 characters standard has been applied here! 👍 😊

0xteet0 commented on src/main/java/oscar/Oscar.java:
```diff
@@ -0,0 +1,68 @@
+import java.util.Scanner;
+
+import oscar.command.Command;
+import oscar.essential.Parser;
```
Nicely done! Imported classes here are listed explicitly! 😊

0xteet0 commented on src/main/java/oscar/command/Command.java:
```diff
@@ -0,0 +1,68 @@
+     * @throws OscarException Failure to validate task number.
+     */
+    public int validateInt(String details, TaskList tasks) throws OscarException {
+        if (details.isEmpty()) {
```
Good to see that single statement conditionals are wrapped by curly brackets still!

0xteet0 commented on src/main/java/oscar/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,80 @@
+     */
+    @Override
+    public void execute(TaskList tasks, Storage storage) throws OscarException {
+        String[] validatedDetails = validate();
```
Great naming here! Abided the convention of plural form to be used on names representing a collection of objects!

0xteet0 commented on src/main/java/oscar/essential/Parser.java:
```diff
@@ -0,0 +1,73 @@
+        case "unmark":
+            return new UnmarkCommand(details);
+
+        // Delete a task if user enters "delete" command
```
Great that comments are indented relative to their position in the code! 😊

0xteet0 commented on src/main/java/oscar/essential/Ui.java:
```diff
@@ -0,0 +1,46 @@
+    /**
+     * Prints divider line in output.
+     */
+    public void showLine() {
```
Is there any reason that you chose to use a method to print the separation line? Perhaps, it might be better to declare it as a global variable to be printed in other methods, avoiding extra calls to the method?

0wl3xSec reviewed (commented):
Overall, the code was well structured, easy to read, with Java coding standards well abided! There are no instances of deep nesting in the code. No common mistakes were made in coding standards! LGTM! 😊😊😊

Cikguseven committed:
Add JavaFX tutorial

Cikguseven committed:
Add new images

Cikguseven committed:
JavaFX WIP

Cikguseven committed:
Add Level 10

Cikguseven committed:
Merge branch 'branch-Level-10'

revdrag commented on src/main/java/oscar/Oscar.java:
```diff
@@ -0,0 +1,56 @@
+ * Chatbot named Oscar that can respond to user input.
+ */
+public class Oscar {
+    static final String FILE_PATH = "./data/tasklist";
```
I like that you use a proper name which is explicit in explaining what the constant mean.

revdrag commented on src/main/java/oscar/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,80 @@
+ * Command to create a new deadline task.
+ */
+public class DeadlineCommand extends Command {
+    private final String details;
```
Maybe it would be better if you used a single-value noun (eg. description) to represent the string as details give the idea that is multi-valued?

revdrag commented on src/main/java/oscar/command/Command.java:
```diff
@@ -0,0 +1,69 @@
+ */
+public abstract class Command {
+    static final DateTimeFormatter DATE_TIME_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
+    private final boolean isExit;
```
I like that you made a boolean sound like one!

revdrag commented on src/main/java/oscar/essential/TaskList.java:
```diff
@@ -0,0 +1,169 @@
+     * @param index Task number.
+     * @return Description of task.
+     */
+    public String delete(int index) {
```
Minor suggestion but perhaps it would be clearer if the method name was deleteTask()?

revdrag reviewed (commented):
Clean and well-documented code! 🤩

Cikguseven committed:
Modify start point of JAR Launcher

JAR launcher was using Oscar.java as the start point before GUI was added.

After adding Javafx for Oscar to have GUI, mainClass.set() in build.gradle needs to have the updated start point launcher file. We use Launcher.java instead of Oscar.java with Javafx.

Cikguseven committed:
Modify JAR launcher

Cikguseven committed:
Merge branch 'master' of https://github.com/Cikguseven/ip

Cikguseven committed:
Add assertions

Code did not contain assertions previously.

Adding assertions to the code help to verify the programme's runtime
state and assumptions.

Assertions have been added liberally as per requirements. Added
assertions are mostly used to ensure that method parameters
are not null.

Cikguseven committed:
Improve code quality

Cikguseven committed:
Merge pull request #1 from Cikguseven/branch-A-CodeQuality

Improve code quality

Cikguseven committed:
Add assertions

Cikguseven committed:
Merge pull request #2 from Cikguseven/branch-A-Assertions

Added assertions to the code to verify the programme's runtime state and assumptions.

Cikguseven committed:
Resolve merge conflict

Cikguseven committed:
Merge branch 'master' of https://github.com/Cikguseven/ip

Cikguseven committed:
Add Note feature

Cikguseven committed:
Merge branch 'branch-D-Notes'

Cikguseven committed:
Improve code quality

Cikguseven committed:
Add user guide

Cikguseven committed:
Update user guide

Cikguseven committed:
Fix checkstyle errors

Cikguseven committed:
Release

Cikguseven committed:
Release v1.0.0

Cikguseven committed:
Update gradle version

Cikguseven committed:
Update README

Cikguseven committed:
Fixed checkstyle

Cikguseven committed:
Set minimum window size

Cikguseven committed:
Update README.md

Updated README to include user guide on how to use Oscar. Relevant commands are explained on how they can be used.
